cpp/include/raft/neighbors/detail/refine_host-inl.hpp

```diff
@@ -43,35 +43,43 @@ template <typename DC, typename IdxT, typename DataT, typename DistanceT, typena
     "neighbors::refine_host(%zu, %zu -> %zu)", n_queries, orig_k, refined_k);
 
   auto suggested_n_threads = std::max(1, std::min(omp_get_num_procs(), omp_get_max_threads()));
-  if (size_t(suggested_n_threads) > n_queries) { suggested_n_threads = n_queries; }
 
-#pragma omp parallel num_threads(suggested_n_threads)
-  {
-    std::vector<std::tuple<DistanceT, IdxT>> refined_pairs(orig_k);
-    for (size_t i = omp_get_thread_num(); i < n_queries; i += omp_get_num_threads()) {
-      // Compute the refined distance using original dataset vectors
+  // For efficiency, each thread should read a certain amount of array elements.
+  // The number of threads for distance computation is determined taking this into account.
+  constexpr int n_elements = 512;
+  size_t max_n_threads = ((n_queries * orig_k * dim) + n_elements - 1) / n_elements;
```

Comment:
Nitpick (requires `#include <raft/util/integer_utils.hpp>`)
```suggestion
  auto max_n_threads = raft::div_rounding_up_safe<size_t>(n_queries * orig_k * dim, n_elements);
```